### PR TITLE
Fix saving of persistent data files on different drives

### LINKED ISSUE #6394: How does pylint decide where it should put stat files?
### Question

I am from the VS Code python extension team, working on building an extension for VS Code using pylint. We have pre-release version out. An issue was reported on the extension that it creates pylint stat files in the project directory. We are not specifically passing anything to change the stat file location (see issue here https://github.com/microsoft/vscode-pylint/issues/30).

Generally, the stat files go into pre-set directory like %USERPROFILE%/AppData/Local/pylint on windows. What can cause these to be written to the current working directory.

We use pass the source code into pylint via stdin, and the only two arguments we use are `--reports=n`, `--output-format=json`. Any clue to debug this would be helpful.

### Documentation for future user

I could not find fall back location for the stat files and when would the fallback location be used.

### Additional context

We use `runpy.run_module` to run pylint, and we send source to pylint via stdin. 
Relevant code here: https://github.com/microsoft/vscode-pylint/blob/725fc7ae415a9638e2a44e922cf5133efde62811/bundled/linter/utils.py#L146-L168

### PR DESCRIPTION
**Pull request by DanielNoord**
- [x] Add a ChangeLog entry describing what your PR does.
- [x] If it's a new feature, or an important bug fix, add a What's New entry in
      `doc/whatsnew/<current release.rst>`.
- [x] Write a good description on what the PR does.
- [x] If you used multiple emails or multiple names when contributing, add your mails
   and preferred name in ``script/.contributors_aliases.json``

## Type of Changes

|     | Type                   |
| --- | ---------------------- |
| ✓   | :bug: Bug fix          |

## Description

Closes #6394 and the underlying problem. See for reference also [Fix saving of persistent data files on different drives](https://github.com/PyCQA/pylint/commit/01ffb60ff4094a0876517f826e54a805cfb50279) and https://github.com/DanielNoord/pylint/pull/135.

Basically drive letters are not handled very well when you concatenate two paths with different drives. We now normalise one of the paths to make sure this is handled better.
Tests show behaviour on both *nix and Windows.

### CODE REVIEW

File: pylint/lint/caching.py [hunk 1]
```diff
@@ -16,7 +16,12 @@
 def _get_pdata_path(
     base_name: Path, recurs: int, pylint_home: Path = Path(PYLINT_HOME)
 ) -> Path:
-    underscored_name = "_".join(str(p) for p in base_name.parts)
```
Comment (Pierre-Sassoulas) [inline pylint/lint/caching.py:19]: It's going to be hard to backport, this function was in ``pylint.config.__init__.py`` in 2.13 and PYLINT_HOME was a global, there was a ``pathlib.Path`` refactor, there was no test previously... I think we should not backport, especially considering that 2.14.0 will be out in such a short time.

**Review by Pierre-Sassoulas — APPROVED** [review]
LGTM, thank you for troubleshooting that @DanielNoord

File: pylint/lint/caching.py (hunk 1 above)
Comment (alkatar21) [reply pylint/lint/caching.py:19]: I think it's ok to release this with 2.14. Especially since I'm probably one of the few who currently has this problem, as long as the extension is not spreaded more. For 2.13.x would probably anyway https://github.com/microsoft/vscode-pylint/issues/30#issuecomment-1118534288 the solution (even if I had not tested under unix)?

Comment (Pierre-Sassoulas) [reply pylint/lint/caching.py:19]: For reference the fix in 2.13 is ``base_name = os.path.splitdrive(base_name.replace(os.sep, "_"))[1]``
